### PR TITLE
feat(cache): add generic cache functions to populate and invalidate cache

### PR DESCRIPTION
**Pull request by dracarys18**
## Type of Change

- [x] Enhancement

## Description

This adds generic helper functions which helps to populate cache and update cache.

## Motivation and Context

So this PR adds two generic functions to help you populate value from redis if it doesn't exist in redis it will get it from DB and insert the value in redis. Usage:

```rust
inhabit_cache(&store,&key, || async{ function_that_inserts_into_db().await}).await
```
And for redact it will update it directly in the DB and remove it from cache

```rust
redact_cache(&store,&key,|| async{function_that_updates_into_db().await}).await
```

## How did you test it?

Tested with #481 PR

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

File: crates/router/src/db/cache.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#![allow(dead_code)]
```
Comment (Narayanbhat166) [inline crates/router/src/db/cache.rs:1]: remove this

File: crates/router/src/db/cache.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+use error_stack::{report, ResultExt};
+
+use super::Store;
+use crate::core::errors::{self, CustomResult};
+
+pub async fn inhabit_cache<T, F, Fut>(
```
Comment (Narayanbhat166) [inline crates/router/src/db/cache.rs:8]: Provide meaningful function name or add comments

**Comment by dracarys18** [comment]
Closing this as this can't be generic because the logic would change for different use cases.

File: crates/router/src/db/cache.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+where
+    T: serde::Serialize + serde::de::DeserializeOwned + std::fmt::Debug,
+    F: FnOnce() -> Fut + Send,
+    Fut: futures::Future<Output = CustomResult<T, errors::StorageError>> + Send,
+{
+    let redis = &store.redis_conn;
+    let redis_val = redis
+        .get_and_deserialize_key::<T>(key, std::any::type_name::<T>())
+        .await;
+    Ok(match redis_val {
```
Comment (jarnura) [inline crates/router/src/db/cache.rs:22]: in OK we can wrap with Ok and we don't want to use ? in line: 34

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**